pkg/controller/endpointslicemirroring/endpointset.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslicemirroring
+
+import (
+	"sort"
+
+	discovery "k8s.io/api/discovery/v1beta1"
+	endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
+)
+
+// endpointHash is used to uniquely identify endpoints. Only including addresses
+// and hostnames as unique identifiers allows us to do more in place updates
+// should attributes such as topology, conditions, or targetRef change.
+type endpointHash string
+type endpointHashObj struct {
+	Addresses []string
+	Hostname  string
+}
+
+func hashEndpoint(endpoint *discovery.Endpoint) endpointHash {
+	sort.Strings(endpoint.Addresses)
+	hashObj := endpointHashObj{Addresses: endpoint.Addresses}
+	if endpoint.Hostname != nil {
+		hashObj.Hostname = *endpoint.Hostname
+	}
+
+	return endpointHash(endpointutil.DeepHashObjectToString(hashObj))
+}
+
+// endpointSet provides simple methods for comparing sets of Endpoints.
+type endpointSet map[endpointHash]*discovery.Endpoint
+
+// Insert adds items to the set.
+func (s endpointSet) Insert(items ...*discovery.Endpoint) endpointSet {
+	for _, item := range items {
+		s[hashEndpoint(item)] = item
+	}
+	return s
+}
+
+// Delete removes all items from the set.
+func (s endpointSet) Delete(items ...*discovery.Endpoint) endpointSet {
+	for _, item := range items {
+		delete(s, hashEndpoint(item))
+	}
+	return s
+}
+
+// Has returns true if and only if item is contained in the set.
+func (s endpointSet) Has(item *discovery.Endpoint) bool {
+	_, contained := s[hashEndpoint(item)]
+	return contained
+}
+
+// Returns an endpoint matching the hash if contained in the set.
+func (s endpointSet) Get(item *discovery.Endpoint) *discovery.Endpoint {
```

Comment:
This is actually a direct copy of the set defined by the EndpointSlice controller. I think that may have been based on another common k8s Set datastructure, but can't quite remember. I'm open to changing the signature of this function at some point, but would prefer to do it once we're able to combine/share some of the code for these controllers. For the sake of simpler backporting I'm trying to avoid any changes like that here.

Reply:
ok